lib/puppet/functions/vault_lookup/lookup.rb

```diff
@@ -2,48 +2,82 @@
   dispatch :lookup do
     param 'String', :path
     optional_param 'String', :vault_url
+    optional_param 'Variant[String,Undef]', :vault_cert_path_segment
```

Comment:
The `optional_param` vs `undef` thing has tripped me up more than once... There's more history in  [PUP-8101](https://tickets.puppetlabs.com/browse/PUP-8101?focusedCommentId=502043&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-502043) and [PUP-11066](https://tickets.puppetlabs.com/browse/PUP-11066?focusedCommentId=827195&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-827195)

My understanding is `optional_param` says I can accept this positional argument or not. But it doesn't specify the data type of the argument. Using `Variant` above will work or you can use a shorter form:

```
optional_param 'Optional[String]', :vault_cert_path_segment
````

This is because the [`undef` data type](https://puppet.com/docs/puppet/7/lang_data_undef.html) is defined so that:

> Several abstract data types can match the undef value ... The Optional type ... matches undef in addition to that type

So this returns `true`:

```
$ bx puppet apply -e 'notice(undef =~ Optional[String])'                                
Notice: Scope(Class[main]): true
```

But this doesn't:

```
$ bx puppet apply -e 'notice(42 =~ Optional[String])'
Notice: Scope(Class[main]): false
```